Introspection code for logging /events data types followup

hackerkid opened the pull request:
Followup of #13204

timabbott commented on templates/zerver/api/events_explore.html:
```diff
@@ -1,10 +1,9 @@
 
 <meta name="viewport" content="width=device-width, initial-scale=1.0">
-{{ bundle('api-explore') }}
-{{ bundle('portico') }}
```
Feel free to squash this commit.

hackerkid commented on zerver/tornado/event_queue.py:
```diff
@@ -1064,11 +1066,23 @@ def send_notification_http(realm: Realm, data: Mapping[str, Any]) -> None:
+        event['instrumented_event_id'] = instrumented_event_id
+
+        copied_event = copy.deepcopy(event)  # type: Dict[str, Any]
+        copied_event['event_description'] = description
```
@timabbott  We donesn't seem to be using value of `event_description` anywhere at the moment? Should I make any changes to display it?

timabbott replied:
I added comments explaining these.  The `event_description` field is not used yet but the intent is it'll let us write in-code documentation that'll appear in these API docs.

hackerkid commented:
@timabbott I have posted a comment in the code. Can you take a look at that?

The test seems to be failing with this error

```bash
Dec 17 11:55:10 pyflakes  | zproject/test_capture_api_events_settings.py:1: '.test_settings.*' imported but unused
```

Should I make pyflakes ignore that line or file or do something else?

I made some changes in the first commit. Documenting it here for easy review.

* renamed LOG_EVENT_TYPES -> LOG_API_EVENT_TYPES
* renamed --capture-events -> --capture-api-events
* Created settings.CAPTURED_API_EVENTS_LOG_PATH instead of hardcoding `static/generated/events_schema_checkers.json`
* Moved some imports to top of the file.
* Removed an unused variable called saved_schema from test_events.py
* Included `events_schema_checkers.json` file.

timabbott commented:
This is great progress.  I did a bit of cleanup work on it and then pushed back to the PR.  A few notes:
* I added comments in several places
* I reordered the commits so that we have a hopefully mergable validators commit first.
* I split out and merged a few prep cleanups ending in 6e75c27f74fbc4b99b6ed88cb299591c04d497ea.
* I added a `coveragerc` entry to deal with `LOG_API_EVENT_TYPES`, because those conditionals will likely not be tested by our unit tests.

A few requests:
* For the `validator.py` commit, we should be using a `TypeVar` in the decorator; the current implementation loses type information.  It's also got some inconsistency about how we declare types (e.g. capitalization of `any`); I'm OK with ignoring that for now since it's primarily for display purposes right now.  So I think we can merge this after fixing the TypeVar thing.
* For the `test-backend` portion commit, we should look at changing `test_settings.py` to pull `LOG_API_EVENT_TYPES` from the environment and having `test-backend` patch the environment; it might be cleaner.  (We use this in a few other places in `test_settings.py` for use by Casper tests, so it's a tested pattern).
* I'd also like to avoid the weird nesting we're ending up with in `test-backend.

timabbott commented:
@hackerkid bump on this.

hackerkid commented on zerver/tornado/event_queue.py:
```diff
@@ -937,6 +939,10 @@ def process_presence_event(event: Mapping[str, Any], users: Iterable[int]) -> No
         presence=event['presence'],
     )
 
+    if "instrumented_event_id" in event:
```
@timabbott Apart from the changes you requested, I made this change as well since the tests were failing.

hackerkid commented:
This is ready for a review.

timabbott commented:
@hackerkid can you rebase this, and also comment on to what extent you address my last batch of feedback?

hackerkid commented:
(I have rebased, but don't review yet)

hackerkid commented:
@timabbott Rebased and addressed your comments.

> For the validator.py commit, we should be using a TypeVar in the decorator; the current implementation loses type information. It's also got some inconsistency about how we declare types (e.g. capitalization of any); I'm OK with ignoring that for now since it's primarily for display purposes right now. So I think we can merge this after fixing the TypeVar thing.

This one is already merged https://github.com/zulip/zulip/commit/1c6435d4cc0b263ba075c3f93618e1f45003d342

> For the test-backend portion commit, we should look at changing test_settings.py to pull LOG_API_EVENT_TYPES from the environment and having test-backend patch the environment; it might be cleaner. (We use this in a few other places in test_settings.py for use by Casper tests, so it's a tested pattern).

Taken care of this.

> I'd also like to avoid the weird nesting we're ending up with in `test-backend.

Taken care of this as well.

timabbott commented:
Thanks for working on this @hackerkid!  Both of us put a good amount of time into this, but we ended up with a better system based on OpenAPI becoming possible a lot sooner than I'd expected, so I think we should probably abandon this plan as no longer necessary.
